nixos: Add 'confinement' options to systemd.services

aszlig opened the pull request:
## Problem statement

Currently, if you want to properly chroot a systemd service, you could do it using `BindReadOnlyPaths=/nix/store` (which is not what I'd call "properly", because the whole store is still accessible) or use a separate derivation that gathers the runtime closure of the service you want to chroot. The former is the easier method and there is also a method directly offered by systemd, called `ProtectSystem`, which still leaves the whole store accessible. The latter however is a bit more involved, because you need to bind-mount each store path of the runtime closure of the service you want to chroot.

This can be achieved using `pkgs.closureInfo` and a small derivation that packs everything into a systemd unit, which later can be added to `systemd.packages`. That's also what I did [several](https://github.com/headcounter/shabitica/blob/3bb01728a0237ad5e7/default.nix#L43-L62) [times](https://github.com/aszlig/avonc/blob/dedf29e092481a33dc/nextcloud.nix#L103-L124) in the past.

However, this process got a bit tedious, so I decided that it would be generally useful for NixOS, so this very implementation was born.

## Description

Now if you want to chroot a systemd service, all you need to do is:

```nix
  {
    systemd.services.yourservice = {
      description = "My Shiny Service";
      wantedBy = [ "multi-user.target" ];

      confinement.enable = true;
      serviceConfig.ExecStart = "${pkgs.myservice}/bin/myservice";
    };
  }
```

If more than the dependencies for the `ExecStart*` and `ExecStop*` (which btw. also includes `script` and `{pre,post}Start`) need to be in the chroot, it can be specified using the `confinement.packages` option. By default (which uses the `full-apivfs`<sup id="note-apivfs">[1](#apivfs)</sup> confinement mode), a user namespace is set up as well and `/proc`, `/sys` and `/dev` are mounted appropriately.

In addition - and by default - a `/bin/sh` executable is provided as well, which is useful for most programs that use the `system()` C library call to execute commands via shell. ~~The shell providing `/bin/sh` is `dash`
instead of the default in NixOS (which is `bash`), because it's way more lightweight and after all we're chrooting because we want to lower the attack surface and it should be only used for `/bin/sh -c something`.~~

Prior to submitting this here, I did a first implementation of this [outside](https://github.com/aszlig/avonc/blob/92a20bece4df54625e/systemd-chroot.nix) of nixpkgs, which duplicated the [pathSafeName](https://github.com/NixOS/nixpkgs/blob/e20ee8a710f3a8ea378bb664c2dbfa32dcf399a7/nixos/modules/system/boot/systemd-lib.nix#L14) functionality from `systemd-lib.nix`, just because it's only a single line.

However, I decided to just re-use the one from systemd here and subsequently made it available when importing `systemd-lib.nix`, so that the `systemd-confinement` implementation also benefits from fixes to that functionality (which is now a proper function).

Unfortunately, we do have a few limitations as well. The first being that `DynamicUser` doesn't work in conjunction with `tmpfs`, because it already sets up a `tmpfs` in a different path and simply ignores the one
we define. We could probably solve this by detecting it and try to bind-mount our paths to that different path whenever `DynamicUser` is enabled.

The second limitation/issue is that `RootDirectoryStartOnly` doesn't work right now, because it only affects the `RootDirectory` option and not the individual bind mounts or our `tmpfs`. It would be helpful if systemd
would have a way to disable specific bind mounts as well or at least have some way to ignore failures for the bind mounts/tmpfs setup.

Another quirk we do have right now is that systemd tries to create a `/usr` directory within the chroot, which subsequently fails. Fortunately, this is just an ugly error and not a hard failure.

##### <b id="apivfs">1</b> The reason this is called `full-apivfs` instead of just `full` is to make room for a *real* `full` confinement mode in the future (once the systemd interface has improved), which is more restrictive even. [↑](#note-apivfs)

aszlig committed:
nixos: Add 'chroot' options to systemd.services

Currently, if you want to properly chroot a systemd service, you could
do it using BindReadOnlyPaths=/nix/store (which is not what I'd call
"properly", because the whole store is still accessible) or use a
separate derivation that gathers the runtime closure of the service you
want to chroot. The former is the easier method and there is also a
method directly offered by systemd, called ProtectSystem, which still
leaves the whole store accessible. The latter however is a bit more
involved, because you need to bind-mount each store path of the runtime
closure of the service you want to chroot.

This can be achieved using pkgs.closureInfo and a small derivation that
packs everything into a systemd unit, which later can be added to
systemd.packages. That's also what I did several times[1][2] in the
past.

However, this process got a bit tedious, so I decided that it would be
generally useful for NixOS, so this very implementation was born.

Now if you want to chroot a systemd service, all you need to do is:

  {
    systemd.services.yourservice = {
      description = "My Shiny Service";
      wantedBy = [ "multi-user.target" ];

      chroot.enable = true;
      serviceConfig.ExecStart = "${pkgs.myservice}/bin/myservice";
    };
  }

If more than the dependencies for the ExecStart* and ExecStop* (which
btw. also includes "script" and {pre,post}Start) need to be in the
chroot, it can be specified using the chroot.packages option. By
default (which uses the "full-apivfs"[3] confinement mode), a user
namespace is set up as well and /proc, /sys and /dev are mounted
appropriately.

In addition - and by default - a /bin/sh executable is provided as well,
which is useful for most programs that use the system() C library call
to execute commands via shell. The shell providing /bin/sh is dash
instead of the default in NixOS (which is bash), because it's way more
lightweight and after all we're chrooting because we want to lower the
attack surface and it should be only used for "/bin/sh -c something".

Prior to submitting this here, I did a first implementation of this
outside[4] of nixpkgs, which duplicated the "pathSafeName" functionality
from systemd-lib.nix, just because it's only a single line.

However, I decided to just re-use the one from systemd here and
subsequently made it available when importing systemd-lib.nix, so that
the systemd-chroot implementation also benefits from fixes to that
functionality (which is now a proper function).

Unfortunately, we do have a few limitations as well. The first being
that DynamicUser doesn't work in conjunction with tmpfs, because it
already sets up a tmpfs in a different path and simply ignores the one
we define. We could probably solve this by detecting it and try to
bind-mount our paths to that different path whenever DynamicUser is
enabled.

The second limitation/issue is that RootDirectoryStartOnly doesn't work
right now, because it only affects the RootDirectory option and not the
individual bind mounts or our tmpfs. It would be helpful if systemd
would have a way to disable specific bind mounts as well or at least
have some way to ignore failures for the bind mounts/tmpfs setup.

Another quirk we do have right now is that systemd tries to create a
/usr directory within the chroot, which subsequently fails. Fortunately,
this is just an ugly error and not a hard failure.

[1]: https://github.com/headcounter/shabitica/blob/3bb01728a0237ad5e7/default.nix#L43-L62
[2]: https://github.com/aszlig/avonc/blob/dedf29e092481a33dc/nextcloud.nix#L103-L124
[3]: The reason this is called "full-apivfs" instead of just "full" is
     to make room for a *real* "full" confinement mode, which is more
     restrictive even.
[4]: https://github.com/aszlig/avonc/blob/92a20bece4df54625e/systemd-chroot.nix

Signed-off-by: aszlig <aszlig@nix.build>

aszlig commented:
Just updated the description a bit to make it less confusing to read :-)

aszlig commented:
Cc: @edolstra, @domenkozar regarding the `pathSafeName` change.

edolstra commented:
Thanks, I've wanted this for a long time. A few comments:

* We should *not* provide a different `/bin/sh` in the chroot, that's just asking for confusion and random shell script breakage. It should be the same shell (i.e. bash) as in a regular environment.

* I don't really like the name "chroot", something like "confine[ment]" or "restrict" seems better. Conceptually we're not providing a completely different filesystem tree but a restricted view of the same tree.

* What's the use case for `chroot.packages`? If a store path doesn't appear in the closure of the exec commands, then how would they even be able to find that path? (BTW we probably should take the closure of the whole unit rather than just the exec commands, to handle things like `Environment` variables).

* We should convert a few services (e.g. PostgreSQL) to use this. (That will also provide some validation.)

aszlig commented:
> We should _not_ provide a different `/bin/sh` in the chroot, that's just asking for confusion and random shell script breakage. It should be the same shell (i.e. bash) as in a regular environment.

As pointed out in the original post, the reason for this is to *only* support the `/bin/sh -c something` case and ideally use an even smaller dummy-shell, that *only* supports that case but nothing else.

> I don't really like the name "chroot", something like "confine[ment]" or "restrict" seems better. Conceptually we're not providing a completely different filesystem tree but a restricted view of the same tree.

Ah, I'm already using `confinement` as a suboption, but I'll change that to `mode` and rename the whole option.

> What's the use case for `chroot.packages`? If a store path doesn't appear in the closure of the exec commands, then how would they even be able to find that path?

This is for getting additional store paths into the chroot, for example like this:

```nix
{
  systemd.services.someservice = {
    environment.CONFIG = someconfig;
    chroot.enable = true;
    chroot.packages = [ someconfig ];
    serviceConfig = { /* ... */ };
  };
}
```

> (BTW we probably should take the closure of the whole unit rather than just the exec commands, to handle things like `Environment` variables).

I *specifically* didn't want to use the whole unit, because it also contains stuff like `coreutils`, which I was trying to avoid during my sandboxing efforts to prevent applications from executing weird stuff I don't want to have.

However, I could implement that by adding another confinement mode that includes the whole unit.

> We should convert a few services (e.g. PostgreSQL) to use this. (That will also provide some validation.)

Unfortunately, PostgreSQL is a bit tricky to convert, because it uses `/tmp` for its socket. We could however bind-mount it to a different directory within the chroot.

aszlig committed:
nixos/systemd-chroot: Rename chroot to confinement

Quoting @edolstra from [1]:

  I don't really like the name "chroot", something like "confine[ment]"
  or "restrict" seems better. Conceptually we're not providing a
  completely different filesystem tree but a restricted view of the same
  tree.

I already used "confinement" as a sub-option and I do agree that
"chroot" sounds a bit too specific (especially because not *only* chroot
is involved).

So this changes the module name and its option to use "confinement"
instead of "chroot" and also renames the "chroot.confinement" to
"confinement.mode".

[1]: https://github.com/NixOS/nixpkgs/pull/57519#issuecomment-472855704

Signed-off-by: aszlig <aszlig@nix.build>

aszlig committed:
nixos/confinement: Allow to configure /bin/sh

Another thing requested by @edolstra in [1]:

  We should not provide a different /bin/sh in the chroot, that's just
  asking for confusion and random shell script breakage. It should be
  the same shell (i.e. bash) as in a regular environment.

While I personally would even go as far to even have a very restricted
shell that is not even a shell and basically *only* allows "/bin/sh -c"
with only *very* minimal parsing of shell syntax, I do agree that people
expect /bin/sh to be bash (or the one configured by environment.binsh)
on NixOS.

So this should make both others and me happy in that I could just use
confinement.binSh = "${pkgs.dash}/bin/dash" for the services I confine.

[1]: https://github.com/NixOS/nixpkgs/pull/57519#issuecomment-472855704

Signed-off-by: aszlig <aszlig@nix.build>

aszlig commented:
@edolstra: Rebased with a few fixups and added the things you mentioned, except the one about adding the full unit closure to the chroot. When looking at how I used this in the past, I frequently use `EnvironmentFile`/`environment` to pass secrets to the service. I wouldn't want those to ever be available to the service in question and as mentioned, there is still the `packages` option to provide extra paths to add to the closure.

I could however add an option which allows to include the full unit closure, but I'm very reluctant to make it the default.

aszlig committed:
nixos/confinement: Allow to include the full unit

From @edolstra at [1]:

  BTW we probably should take the closure of the whole unit rather than
  just the exec commands, to handle things like Environment variables.

With this commit, there is now a "fullUnit" option, which can be enabled
to include the full closure of the service unit into the chroot.

However, I did not enable this by default, because I do disagree here
and *especially* things like environment variables or environment files
shouldn't be in the closure of the chroot.

For example if you have something like:

  { pkgs, ... }:

  {
    systemd.services.foobar = {
      serviceConfig.EnvironmentFile = ${pkgs.writeText "secrets" ''
        user=admin
        password=abcdefg
      '';
    };
  }

We really do not want the *file* to end up in the chroot, but rather
just the environment variables to be exported.

Another thing is that this makes it less predictable what actually will
end up in the chroot, because we have a "globalEnvironment" option that
will get merged in as well, so users adding stuff to that option will
also make it available in confined units.

I also added a big fat warning about that in the description of the
fullUnit option.

[1]: https://github.com/NixOS/nixpkgs/pull/57519#issuecomment-472855704

Signed-off-by: aszlig <aszlig@nix.build>

aszlig commented:
> I could however add an option which allows to include the full unit closure, but I'm very reluctant to make it the default.

Added.

edolstra commented:
Well, if you think that `fullUnit` is a bad idea, then maybe it's better not to have it.

Infinisil commented on nixos/modules/security/systemd-confinement.nix:
```diff
@@ -0,0 +1,192 @@
+          MountFlags = lib.mkDefault "private";
+        } // lib.optionalAttrs (config.confinement.mode == "full-apivfs") {
+          MountAPIVFS = true;
+          PrivateDevices = true;
```
Considering that at some point https://github.com/NixOS/nixpkgs/issues/14645 might progress (I intend to give it a shot), it could be a good idea to not rely on these options all being `false` by default.

aszlig replied:
What's the actual problem here? If they're enabled by default (given the default priority of course), the option merge will not conflict.

Infinisil replied:
If the default for them is `true`, then `mode = "chroot-only"` will still have these options all set to `true` and not `false`.

aszlig replied:
Okay, got it, I thought you meant this was for `full-apivfs`, will change that...

Infinisil commented on nixos/modules/security/systemd-confinement.nix:
```diff
@@ -0,0 +1,192 @@
+      };
+
+      options.confinement.mode = lib.mkOption {
+        type = types.enum [ "full-apivfs" "chroot-only" ];
```
What's this option useful for? If people want to turn off certain isolation options, they can set `PrivateDevices = false`, etc. themselves. This option would only introduce one additional layer of abstraction, of which NixOS has enough already imo

aszlig replied:
This option is actually very useful, particularly the `chroot-only` mode, which allows the application to take care of further restricting itself on its own. Not needing to have `/dev`, `/proc` and `/sys` lowers the attack surface by a huge factor, but of course the application needs to have sandboxing built-in (LibreOffice Online does that for example).

Infinisil replied:
Yeah, but this all is still possible if we remove this option and have a `full-apivfs` setup by default (setting all the `serviceConfig` sandboxy values to `true` with a `mkDefault`), and then let the service writers turn off every protection they don't want manually.

aszlig replied:
I beg to differ here, because the unit options picked by `full-apivfs` unfortunately quite version-specific to systemd. For example I would rather _avoid_ the need to use the `Protect*` options, because they actually increase the attack surface (because they use a blacklist-approach) but right now there is no better option as you just can enable APIVFS or disable it, rather than specifying which pseudo-fs to mount. Additionally, I also want to get in PID namespacing as soon as systemd provides an option.

That's why the `chroot-only` option gets you to the point where you don't need to worry about future systemd updates. Let's say you only want a chroot without anything else and PID namespacing gets enabled, you need to keep track of that and disable it accordingly (after you're done debugging why your service suddenly fails).

Infinisil replied:
Alright this sounds more reasonable then.

Infinisil commented on nixos/modules/security/systemd-confinement.nix:
```diff
@@ -0,0 +1,192 @@
+      };
+
+      options.confinement.binSh = lib.mkOption {
+        type = types.nullOr types.path;
```
Also for this option I'm not sure if it's really good to provide it. Do you have any usecase that absolutely requires `dash` or so? I think it would be fine to just hardcode to always use `${pkgs.bash}/bin/sh` here

aszlig replied:
Not particularly `dash` but an even more restrictive shell that only allows `/bin/sh -c foo` and only does minimal parsing and no I/O redirection. I use that for my own sandboxed deployments a lot (although it's application-specific, still searching for a better generic solution, some minimal "fake-shell" that can be configured according to some rules), plus even going with dash also has the advantage of having a huge speed boost and it's way more minimal, consider:

```
sh -c 'for i in $(seq 1000); do bash -c :; done'  2.07s user 1.23s system 100% cpu 3.281 total
sh -c 'for i in $(seq 1000); do dash -c :; done'  0.27s user 0.70s system 103% cpu 0.937 total
```

infinisil reviewed (changes requested):
Also agree with @edolstra on fullUnit, `packages` should be good enough in those cases where `ExecStart` doesn't reference what you need.

We can always add more options later if they turn out to be needed.

danbst commented on nixos/modules/security/systemd-confinement.nix:
```diff
@@ -0,0 +1,192 @@
+
+          <note><para>Also, the store paths listed in <option>path</option> are
+          <emphasis role="strong">not</emphasis> included in the closure as
+          well as paths from other options except those listed
```
I don't understand why not. I place into `path` things, that are queries by service in runtime, exactly because those are not present in closure of ExecStart.

Agree about other variables.

aszlig replied:
@danbst: It's not only about `path` but also environment and other variants, that's why I introduced the `fullUnit` option, so it's possible to shovel more paths into the closure if you *really* want to. But after all the goal is to minimize the available paths. Including `path` there, would also mean that we get coreutils/findutils/grep/sed and systemd for *every* service using confinement plus the one you additionally specify, unless you `mkOverride` with a lower prio.

danbst replied:
gotcha. Didn't know `path` is preloaded with toolings.

danbst commented:
Maybe I'm a bit late, but what about reuse `sanbox` word instead of `confinement`?

aszlig commented:
@danbst: I thought about `sandbox` in the first place, but for a full sandbox I think this is not exhaustive enough because it only confines on the file system level. That's why I also went for `chroot` in my original implementation.

aszlig committed:
nixos/confinement: Explicitly set serviceConfig

My implementation was relying on PrivateDevices, PrivateTmp,
PrivateUsers and others to be false by default if chroot-only mode is
used.

However there is an ongoing effort[1] to change these defaults, which
then will actually increase the attack surface in chroot-only mode,
because it is expected that there is no /dev, /sys or /proc.

If for example PrivateDevices is enabled by default, there suddenly will
be a mounted /dev in the chroot and we wouldn't detect it.

Fortunately, our tests cover that, but I'm preparing for this anyway so
that we have a smoother transition without the need to fix our
implementation again.

Thanks to @Infinisil for the heads-up.

[1]: https://github.com/NixOS/nixpkgs/issues/14645

Signed-off-by: aszlig <aszlig@nix.build>

aszlig committed:
nixos/confinement: Remove handling for StartOnly

Noted by @Infinisil on IRC:

   infinisil: Question regarding the confinement PR
   infinisil: On line 136 you do different things depending on
              RootDirectoryStartOnly
   infinisil: But on line 157 you have an assertion that disallows that
              option being true
   infinisil: Is there a reason behind this or am I missing something

I originally left this in so that once systemd supports that, we can
just flip a switch and remove the assertion and thus support
RootDirectoryStartOnly for our confinement module.

However, this doesn't seem to be on the roadmap for systemd in the
foreseeable future, so I'll just remove this, especially because it's
very easy to add it again, once it is supported.

Signed-off-by: aszlig <aszlig@nix.build>

aszlig committed:
nixos/confinement: Use PrivateMounts option

So far we had MountFlags = "private", but as @Infinisil has correctly
noticed, there is a dedicated PrivateMounts option, which does exactly
that and is better integrated than providing raw mount flags.

When checking for the reason why I used MountFlags instead of
PrivateMounts, I found that at the time I wrote the initial version of
this module (Mar 12 06:15:58 2018 +0100) the PrivateMounts option didn't
exist yet and has been added to systemd in Jun 13 08:20:18 2018 +0200.

Signed-off-by: aszlig <aszlig@nix.build>

infinisil reviewed (approved):
I gave some review over IRC, which @aszlig addressed with the last two commits. This PR looks good to me now and I'd like to merge this soon.

aszlig commented:
I also added the `needs: changelog` label, because it should make people more aware of this when writing NixOS modules and we ideally have more services using this option.

Just also checking back with @samueldr and @lheckemann whether we possibly want this in NixOS 19.03.

samueldr commented:
:+1: for backporting as this only enhances the feature set of nixos, and does not (at a glance) change existing behaviours.

aszlig committed:
nixos/release-notes: Add entry about confinement

First of all, the reason I added this to the "highlights" section is
that we want users to be aware of these options, because in the end we
really want to decrease the attack surface of NixOS services and this is
a step towards improving that situation.

The reason why I'm adding this to the changelog of the NixOS 19.03
release instead of 19.09 is that it makes backporting services that use
these options easier. Doing the backport of the confinement module after
the official release would mean that it's not part of the release
announcement and potentially could fall under the radar of most users.

These options and the whole module also do not change anything in
existing services or affect other modules, so they're purely optional.

Adding this "last minute" to the 19.03 release doesn't hurt and is
probably a good preparation for the next months where we hopefully
confine as much services as we can :-)

I also have asked @samueldr and @lheckemann, whether they're okay with
the inclusion in 19.03. While so far only @samueldr has accepted the
change, we can still move the changelog entry to the NixOS 19.09 release
notes in case @lheckemann rejects it.

Signed-off-by: aszlig <aszlig@nix.build>

lheckemann commented:
I'm neutral on backporting. On the one hand, as @samueldr says, this doesn't change any existing functionality. On the other hand, it's a new feature and we're well past the feature freeze, which AFAIU is intended so that new, potentially buggy features aren't introduced in the "stabilisation period". It *is* a cool feature though… :)

arianvp commented:
I don't fully understand why this doesn't interact well with `DynamicUser`. I make heavy use of it and it would be awesome to use it together with this feature.  Could you elaborate a bit?

aszlig commented:
@arianvp: See ac64ce994509aaad8c5b55254595a5f989ba24e9:

> Unfortunately, we do have a few limitations as well. The first being
> that DynamicUser doesn't work in conjunction with tmpfs, because it
> already sets up a tmpfs in a different path and simply ignores the one
> we define. We could probably solve this by detecting it and try to
> bind-mount our paths to that different path whenever DynamicUser is
> enabled.

aszlig commented:
@arianvp: Okay, this is getting a [bit more complicated](https://gist.github.com/aszlig/11b65f5dba569a166642f4df9c216401), because systemd doesn't seem to take the `TemporaryFilesystem` into account, so we need to create a separate mount unit for working around this behaviour. I'd like to do this in a followup PR though (and maybe even fix this in systemd directly), because this needs a lot more tests to make sure it stays consistent with the current implementation without `DynamicUser = true`.

Infinisil commented:
I would've squashed a bit, but oh well

hyperfekt commented:
[`DynamicUser=true` implies `ProtectSystem`, which overrides `TemporaryFilesystem`.](https://github.com/systemd/systemd/issues/18999#issuecomment-798919429)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixos-policy-regarding-systemd-confinement/18976/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/hardening-systemd-services/17147/18
